### ISSUE TITLE
metric problems in xgboost rank

### ISSUE THREAD

**Issue by zhangy11** [dmlc/xgboost#2208]
I use xgboost to solve rank problem. The relevance values of items are binary {0, 1}. For each query, only one item is 1, the others are all 0.
The params are set as follows:
`param = {
        'max_depth':3, 
        'eta':0.1, 
        'silent':0, 
        'objective':'rank:pairwise', 
    }`
`param['nthread'] = 16`
`param['eval_metric'] = ['map@1-','map@1','map@5-','map@5','map@1000','map@1000-','ndcg@1-','ndcg@1','ndcg@5-','ndcg@5','ndcg@1000-','ndcg@1000']` 
`evallist  = [(dtrain, 'train'), (dtest, 'test')]`

After training, the eval_metric are as follows:
`train-map@1-:0.955956	train-map@1:0.955956	train-map@5-:0.976864	train-map@5:0.976864	train-map@1000:0.976864	train-map@1000-:0.976864	train-ndcg@1-:0.955956	train-ndcg@1:0.955956	train-ndcg@5-:0.982867	train-ndcg@5:0.982867	train-ndcg@1000-:0.982867	train-ndcg@1000:0.982867`	
`test-map@1-:0.766555	test-map@1:0.766555	test-map@5-:0.838688	test-map@5:0.838688	test-map@1000:0.844153	test-map@1000-:0.844153	test-ndcg@1-:0.765355	test-ndcg@1:0.765355	test-ndcg@5-:0.864858	test-ndcg@5:0.864858	test-ndcg@1000-:0.880889	test-ndcg@1000:0.880889`

I have seen the wiki of map and ndcg. I'm confused with this result very much. In parameter.md, it says:
"ndcg-","map-","ndcg@n-","map@n-": In XGBoost, NDCG and MAP will evaluate the score of a list without any positive samples as 1. By adding "-" in the evaluation metric XGBoost will evaluate these score as 0 to be consistent under some conditions.
**Why are they equal whether I add "-" after map@n and ndcg@n or not.**

**Comment by tqchen** [comment]
They are only used in strange cases where the entire list is not relevant, so if your dataset is well formed, there won't be difference.

**Comment by pommedeterresautee** [comment]
It's only when you have no positive result in a group. Depending of the paper/library, different behavior are applied. This is the XGBoost way to do it.
